presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownSpatialProjections.java

```diff
@@ -0,0 +1,207 @@
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        // TODO Consider applying this rule for distributed joins as well
+        if (isDistributedJoinEnabled(context.getSession())) {
+            return Result.empty();
+        }
+
+        JoinNode joinNode = captures.get(JOIN);
+
+        Expression filter = node.getPredicate();
+        List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filter);
+        for (FunctionCall spatialFunction : spatialFunctions) {
+            List<Expression> arguments = spatialFunction.getArguments();
+            Verify.verify(arguments.size() == 2);
+
+            Expression firstArgument = arguments.get(0);
+            Expression secondArgument = arguments.get(1);
+
+            Set<String> firstSymbols = getSymbolNames(extractUnique(firstArgument));
+            Set<String> secondSymbols = getSymbolNames(extractUnique(secondArgument));
+
+            if (firstSymbols.isEmpty() || secondSymbols.isEmpty()) {
+                return Result.empty();
+            }
+
+            Optional<Symbol> newFirstSymbol = newSymbol(context, firstArgument);
+            Optional<Symbol> newSecondSymbol = newSymbol(context, secondArgument);
+
+            Expression newFirstArgument = toExpression(newFirstSymbol, firstArgument);
+            Expression newSecondArgument = toExpression(newSecondSymbol, secondArgument);
+
+            Expression newSpatialFunction = new FunctionCall(spatialFunction.getName(), ImmutableList.of(newFirstArgument, newSecondArgument));
+            Expression newFilter = replaceExpression(filter, ImmutableMap.of(spatialFunction, newSpatialFunction));
+
+            PlanNode leftNode = joinNode.getLeft();
+            PlanNode rightNode = joinNode.getRight();
+
+            Set<String> leftSymbols = getSymbolNames(leftNode.getOutputSymbols());
+            Set<String> rightSymbols = getSymbolNames(rightNode.getOutputSymbols());
+
+            PlanNode newLeftNode;
+            PlanNode newRightNode;
+
+            if (leftSymbols.containsAll(firstSymbols)
+                    && containsNone(leftSymbols, secondSymbols)
+                    && rightSymbols.containsAll(secondSymbols)
+                    && containsNone(rightSymbols, firstSymbols)) {
+                newLeftNode = newFirstSymbol.map(symbol -> addProjection(context, leftNode, symbol, firstArgument)).orElse(leftNode);
+                newRightNode = newSecondSymbol.map(symbol -> addProjection(context, rightNode, symbol, secondArgument)).orElse(rightNode);
+            }
+            else if (leftSymbols.containsAll(secondSymbols)
+                    && containsNone(leftSymbols, firstSymbols)
+                    && rightSymbols.containsAll(firstSymbols)
+                    && containsNone(rightSymbols, secondSymbols)) {
+                newLeftNode = newSecondSymbol.map(symbol -> addProjection(context, leftNode, symbol, secondArgument)).orElse(leftNode);
```

Comment:
what if leftNode has other symbols?

Reply:
addProjection method preserves all the output symbols of the node and adds one more:

```
private PlanNode addProjection(Context context, PlanNode node, Symbol symbol, Expression expression)
    {
        Assignments.Builder projections = Assignments.builder();
        for (Symbol outputSymbol : node.getOutputSymbols()) {
            projections.putIdentity(outputSymbol);
        }

        projections.put(symbol, expression);
        return new ProjectNode(context.getIdAllocator().getNextId(), node, projections.build());
    }
```